Ios yaml (#49)
codemagic-ci-cd/codemagic-docs: content/building/yaml.md

          vars:
            PUBLIC_ENV_VAR: "value here"
          flutter: stable
          xcode: latest
        cache:
          cache_paths:

>             - ~/.pub

Reviewer: Please add `~.pub-cache` instead of `.pub`.

Reviewer 2: we can't cache `$FCI_BUILD_DIR/build` since it causes compilation errors

Reviewer: ```suggestion
            - ~/.pub-cache
```

        triggering:
        branch_patterns:
        scripts:
        artifacts:
        publishing:

The main sections in each workflow are described below.

### Environment

> `environment:` Contains your environment variables and enables to specify the version of Flutter used for building. This is also where you are required to add credentials and API keys required for code signing. Make sure to [encrypt the values](#encrypting-sensitive-data) of variables that hold sensitive data. 

Reviewer: > This is also where you are required to add credentials and API keys required for code signing.

It is not _required_ to add the credentials here. Each property can be specified as a command argument to programs with dedicated flags.

As an example the following yield exactly the same result:
```bash
app-store-connect fetch‑signing‑files --issuer-id=<ISSUER-ID> --key-id=<KEY-IDENTIFIER> --private-key=@file:/path/to/app-store-connect-private-key.p8 --certificate-key=@file:/path/to/certificate-private-key.pem
``` 
and
```bash
APP_STORE_CONNECT_ISSUER_ID=<ISSUER-ID>
APP_STORE_CONNECT_KEY_IDENTIFIER=<KEY-IDENTIFIER>
APP_STORE_CONNECT_PRIVATE_KEY=<app-store-connect-private-key-contents-as-pem>
CERTIFICATE_PRIVATE_KEY=<code-signing-certificate-private-key-contents-as-pem>

app-store-connect fetch‑signing‑files
``` 

The environment variables are just shortcuts / default fallbacks to avoid cluttering the script phases. There are number of ways those values can be specified as stated [here](https://github.com/codemagic-ci-cd/cli-tools/blob/master/docs/app-store-connect/fetch%E2%80%91signing%E2%80%91files.md#--issuer-idissuer_id).

        
        # iOS automatic code signing
        APP_STORE_CONNECT_ISSUER_ID: Encrypted(...)
        APP_STORE_CONNECT_KEY_IDENTIFIER: Encrypted(...)
        APP_STORE_CONNECT_PRIVATE_KEY: Encrypted(...)
        CERTIFICATE_PRIVATE_KEY: Encrypted(...)

        # iOS manual code signing
        CM_CERTIFICATE: Encrypted(...)
        CM_CERTIFICATE_PASSWORD: Encrypted(...)
        CM_PROVISIONING_PROFILE: Encrypted(...)

      flutter: stable   # Define the channel name or version (e.g. v1.13.4)
      xcode: latest     # Define latest, edge or version (e.g. 11.2)

#### Setting up code signing for iOS

In order to use **automatic code signing** where Codemagic creates and manages signing certificates and provisioning profiles on your behalf, you need to configure API access to App Store Connect and define the following environment variables: 

* `APP_STORE_CONNECT_PRIVATE_KEY`

  It is recommended to create a dedicated App Store Connect API key for Codemagic in [App Store Connect](https://appstoreconnect.apple.com/access/api).

  1. Log in to App Store Connect and navigate to **Users and Access > Keys**.
  2. Click on the '+' sign to generate a new API key. 

>   3. Enter the name for the key and select an access level (`Admin`, `App Manager` or `Developer`).

Reviewer: I'm not sure if `App Manager` access level will make the cut.

  4. Click **Generate**.
  5. As soon as the key is generated, you can see it added in the list of active keys. Click **Download API Key** to save the private key. Note that the key can only be downloaded once.

* `APP_STORE_CONNECT_KEY_IDENTIFIER`

  In **App Store Connect > Users and Access > Keys**, this is the **Key ID** of the key.

* `APP_STORE_CONNECT_ISSUER_ID`

  In **App Store Connect > Users and Access > Keys**, this is the **Issuer ID** displayed above the table of active keys.

* `CERTIFICATE_PRIVATE_KEY`

  RSA 2048 bit private key to be included in the signing certificate. Read more about it [here](https://help.apple.com/xcode/mac/current/#/dev1c7c2c67d).

> In order to use **manual code signing**, upload the encrypted signing certificate, the certificate password (if the certificate is password-protected) and the provisioning profile to the following environment variables:

Reviewer: ```suggestion
In order to use **manual code signing**, [encrypt](https://docs.codemagic.io/building/yaml/#encrypting-sensitive-data) your signing certificate, the certificate password (if the certificate is password-protected) and the provisioning profile, and set the encrypted values to the following environment variables:
```

**Note on building Android app bundles**

If your app settings in Codemagic have building Android app bundles enabled, we will automatically include a script for generating a signed `app-universal.apk` during the YAML export. If you're creating a YAML file from a scratch, add the script below to receive this file:

>     # fetch codemagic helper scripts

Reviewer: This block is obsolete and should be replaced. I'll provide correct version soon.

Reviewer: The following lines should be removed
```
# fetch codemagic helper scripts
...
~/codemagic-build-scripts/android/generate-universal-apks \
  ...
  --pattern 'build/**/outputs/**/*.aab'
```

And replaced with
```
# generate signed universal apk with user specified keys
universal-apk generate \
      --ks /tmp/keystore.keystore \
      --ks-pass $CM_KEYSTORE_PASSWORD \
      --ks-key-alias $CM_KEY_ALIAS_USERNAME \
      --key-pass $CM_KEY_ALIAS_PASSWORD \
      --pattern 'build/**/outputs/**/*.aab'
```

      --key-pass $CM_KEY_ALIAS_PASSWORD \
      --pattern 'build/**/outputs/**/*.aab'
  
#### Building for iOS

{{<notebox>}}
Codemagic uses the [app-store-connect](https://github.com/codemagic-ci-cd/cli-tools/blob/master/docs/app-store-connect/README.md) utility for generating and managing certificates and provisioning profiles and performing code signing.
{{</notebox>}}

Below is an example of building a Flutter app for iOS with automatic code signing. 

    scripts:
          - flutter packages pub get
          - flutter analyze
          - flutter test
          - find . -name "Podfile" -execdir pod install \;
          - keychain initialize

>           - app-store-connect fetch-signing-files "com.example.capybara.dev" --type IOS_APP_DEVELOPMENT --create       # Specify bundle ID and provisioning profile type

Reviewer: ```suggestion
          - app-store-connect fetch-signing-files "io.codemagic.app" \    # Specify bundle ID
                --type IOS_APP_DEVELOPMENT \                                             # Specify provisioning profile type
                --create                                                                                        # Allow creating new profiles and
                                                                                                                      # certificates if existing resources are not found.
```

          - keychain add-certificates
          - flutter build ios --debug --flavor dev --no-codesign
          - xcode-project use-profiles

>           - xcode-project build-ipa --workspace ios/Runner.xcworkspace --config Debug --scheme dev

Reviewer: ```suggestion
          - xcode-project build-ipa --workspace ios/Runner.xcworkspace --scheme dev
```

Below is an example of building a Flutter app for iOS with manual code signing.

    scripts:
      - flutter packages pub get
      - flutter analyze
      - flutter test
      - find . -name "Podfile" -execdir pod install \;
      - keychain initialize
      - |
        # set up provisioning profiles
        PROFILES_HOME="$HOME/Library/MobileDevice/Provisioning Profiles"
        mkdir -p "$PROFILES_HOME"
        PROFILE_PATH="$(mktemp "$PROFILES_HOME"/$(uuidgen).mobileprovision)"
        echo ${CM_PROVISIONING_PROFILE} | base64 --decode > $PROFILE_PATH
        echo "Saved provisioning profile $PROFILE_PATH"
      - |
        # set up signing certificate
        echo $CM_CERTIFICATE | base64 --decode > /tmp/certificate.p12

        # when using a password-protected certificate
        keychain add-certificates --certificate /tmp/certificate.p12 --certificate-password $CM_CERTIFICATE_PASSWORD
        # when using a certificate that is not password-protected
        keychain add-certificates --certificate /tmp/certificate.p12

      - flutter build ios --debug --flavor dev --no-codesign
      - xcode-project use-profiles

>       - xcode-project build-ipa --workspace ios/Runner.xcworkspace --config Debug --scheme dev

Reviewer: ```suggestion
      - xcode-project build-ipa --workspace ios/Runner.xcworkspace --scheme Runner
```
